renderdoc/driver/d3d12/d3d12_shader_feedback.cpp

```diff
@@ -1218,6 +1569,55 @@ void D3D12Replay::FetchShaderFeedback(uint32_t eventId)
         }
       }
 
+      ShaderStage shaderStages[] = {ShaderStage::Compute,  ShaderStage::Vertex,
+                                    ShaderStage::Hull,     ShaderStage::Domain,
+                                    ShaderStage::Geometry, ShaderStage::Pixel};
+      DXBCBytecode::OperandType operandTypes[] = {DXBCBytecode::TYPE_RESOURCE,
+                                                  DXBCBytecode::TYPE_SAMPLER};
+      uint32_t maxDescriptorsPerType[] = {maxDescriptors, maxSamplerDescriptors};
+      for(int operandIndex = 0; operandIndex < 2; ++operandIndex)
+      {
+        D3D12FeedbackKey directAccessKey = GetDirectHeapAccesKey(operandTypes[operandIndex]);
+        for(int shaderStage = 0; shaderStage < 6; shaderStage++)
+        {
+          if(slots[shaderStage].find(directAccessKey) == slots[shaderStage].end())
+            continue;
+          D3D12FeedbackSlot &feedbackSlots = slots[shaderStage].at(directAccessKey);
+          for(uint32_t i = feedbackSlots.Slot();
+              i < feedbackSlots.Slot() + maxDescriptorsPerType[operandIndex]; ++i)
+          {
+            if(slotsData[i] == magicFeedbackValue)
+            {
+              uint32_t usedSlot = i - feedbackSlots.Slot();
+              D3D12FeedbackBindIdentifier directAccessIdentifier = {};
+              directAccessIdentifier.descIndex = usedSlot;
+              directAccessIdentifier.rootEl = ~0U;
+              directAccessIdentifier.rangeIndex = ~0U;
+              directAccessIdentifier.directAccess = true;
+              if(operandTypes[operandIndex] == DXBCBytecode::TYPE_SAMPLER)
+              {
+                directAccessIdentifier.bindType = BindType::Sampler;
+              }
+              else
+              {
+                RDCASSERTNOTEQUAL(resourceHeap, NULL);
+                D3D12Descriptor *desc =
+                    (D3D12Descriptor *)resourceHeap->GetCPUDescriptorHandleForHeapStart().ptr;
+                desc += usedSlot;
+                if(desc->GetType() == D3D12DescriptorType::CBV)
+                  directAccessIdentifier.bindType = BindType::ConstantBuffer;
+                if(desc->GetType() == D3D12DescriptorType::SRV)
+                  directAccessIdentifier.bindType = BindType::ReadOnlyResource;
+                if(desc->GetType() == D3D12DescriptorType::UAV)
+                  directAccessIdentifier.bindType = BindType::ReadWriteResource;
```

Comment:
It is indeed to get the resource identifiers properly sorted.
I originally started by using ``dx.op.annotateHandle`` but as you said the lack of documentation made me fear that I could miss some cases, however from my limited testing the instruction seems to be there even compiled as -O3 without debug informations.

I will put it back instead of looking at the heap, though rather than using bits in the feeback value I'd rather allocate a range in the feedback UAV per resource type ( [SRV][UAV][CBV][SAMPLER] instead of current [SRV,UAV,CBV][SAMPLER] ranges ), it may use more memory but since the ideal case would be a range per ``CreateHandleFromHeap`` to correlate resource use with shader source, I think it it non-significant.

Reply:
Ok I'll change it to use bits in the magic value.
Would you also merge the feedback range for samplers with the range for SRV/CBV/UAV ?
So that an index in the feedback range could represent at the same a descriptor in the sampler heap and one of SRV/CBV/UAV descriptor in the resource heap